harness/determined/profiler.py

```diff
@@ -0,0 +1,559 @@
+                        to_post_format(
+                            self.batch[metric_name][gpu_uuid],
+                            make_labels(
+                                metric_name, SYSTEM_METRIC_TYPE_ENUM, gpu_uuid_label=gpu_uuid
+                            ),
+                        )
+                    )
+
+        return trial_profiler_metrics_batches
+
+
+class StartMessage:
+    pass
+
+
+class ShutdownMessage:
+    pass
+
+
+class SysMetricCollectorThread(threading.Thread):
+    """
+    Background thread for collecting profiler metrics at a high granularity and shipping them to
+    the master
+
+    - SimpleCpuUtilization = Measured in percent
+    - FreeMemory = Measured in Gigabytes
+    - NetworkSentThroughput = Measured in Gigabit/s
+    - NetworkRecvThroughput = Measured in Gigabit/s
+    - DiskIops
+    - DiskReadThroughput = Measured in bytes/second
+    - DiskWriteThroughput = Measured in bytes/second
+    - GpuUtilization = Measured in percent
+    """
+
+    FLUSH_INTERVAL = 10  # How often to make API calls
+    MEASUREMENT_INTERVAL = 0.1
+
+    def __init__(self, trial_id: int, agent_id: str, send_queue: queue.Queue):
+
+        self.current_batch_idx = 0
+        self.send_queue = send_queue
+        self.control_queue: "queue.Queue[Union['StartMessage', 'ShutdownMessage']]" = queue.Queue()
+        self.current_batch = SysMetricBatcher(trial_id, agent_id)
+        self.current_batch.clear()
+
+        super().__init__(daemon=True)
+
+    def activate(self) -> None:
+        self.control_queue.put(StartMessage())
+
+    def kill(self) -> None:
+        self.control_queue.put(ShutdownMessage())
+
+    def update_batch_idx(self, new_batch_idx: int) -> None:
+        self.current_batch_idx = new_batch_idx
+
+    def run(self) -> None:
+        last_measurement_time = None
+        batch_start_time = None
+        cpu_util_collector = SimpleCpuUtilCollector()
+        gpu_util_collector = GpuUtilCollector()
+        gpu_memory_collection = GpuMemoryCollector()
+        network_throughput_collector = NetThroughputCollector()
+        free_memory_collector = FreeMemoryCollector()
+        disk_collector = DiskReadWriteRateCollector()
+
+        msg = self.control_queue.get()
+        if isinstance(msg, ShutdownMessage):
+            return
+
+        batch_start_time = time.time()
+        next_collection = time.time()
+
+        while True:
+            now = time.time()
+            if now < next_collection:
+                # check for quit message while we wait
+                sleep_time = next_collection - now
+                try:
+                    msg = self.control_queue.get(timeout=sleep_time)
+                    if isinstance(msg, ShutdownMessage):
+                        # Drop any partial batches
+                        return
+                except queue.Empty:
+                    pass
+
+            next_collection += self.MEASUREMENT_INTERVAL
```

Comment:
q: should this be
```suggestion
            next_collection = time.now() + self.MEASUREMENT_INTERVAL
```
in case `MEASUREMENT_INTERVAL` is very small?

 In practice, `next_collection - now` could be negative.

Reply:
This was my idea actually.  This keeps the interval as close to regular as possible for normal intervals.  If the interval is too small, it exhibits the behavior of collecting as fast as technically possible, which seems like the right thing to do if the interval is too small.

Reply:
If sleep time can be negative, it seems fine to me.

Reply:
If the timeout is negative, the code will raise an exception. Good catch. I changed it so that the sleep time is never less than 0.

Reply:
Fixed